reactor-core/src/test/java/reactor/core/publisher/SinksTest.java

```diff
@@ -0,0 +1,721 @@
+					          .with().pollDelay(1, TimeUnit.SECONDS)
+					          .untilAsserted(() -> test1.assertValueCount(2));
+				}
+				finally {
+					test1.cancel();
+				}
+			});
+			final Future<?> f2 = es.submit(() -> {
+				AssertSubscriber<Integer> test2 = AssertSubscriber.create(1);
+				flux.subscribe(test2);
+				requestLatch.countDown();
+
+				test2.awaitAndAssertNextValues(1);
+				try {
+					Awaitility.await().atMost(2, TimeUnit.SECONDS)
+					          .with().pollDelay(1, TimeUnit.SECONDS)
+					          .untilAsserted(() -> test2.assertValueCount(1));
+				}
+				finally {
+					test2.cancel();
+				}
+			});
+
+			requestLatch.await(1, TimeUnit.SECONDS);
+			sink.next(1)
+			    .next(2)
+			    .next(3)
+			    .next(4)
+			    .complete();
+
+			f1.get();
+			f2.get();
+		}
+
+		@Test
+		void doesNotReplayToLateSubscribers() {
+			AssertSubscriber<Integer> s1 = AssertSubscriber.create();
+			AssertSubscriber<Integer> s2 = AssertSubscriber.create();
+
+			flux.subscribe(s1);
+			sink.next(1).next(2).next(3);
+			s1.assertValues(1, 2, 3);
+
+			flux.subscribe(s2);
+			s2.assertNoValues().assertNotComplete();
+
+			sink.complete();
+			s1.assertValueCount(3).assertComplete();
+			s2.assertNoValues().assertComplete();
+		}
+
+		@Test
+		public void doesNotBufferBeforeFirstSubscriber() {
+			AssertSubscriber<Integer> first = AssertSubscriber.create();
+
+			sink.next(1).next(2).next(3);
+			flux.subscribe(first);
+
+			first.assertNoValues().assertNotComplete();
+
+			sink.complete();
+			first.assertNoValues().assertComplete();
+		}
+
+		@Test
+		public void immediatelyCompleteFirstSubscriber() {
+			AssertSubscriber<Integer> first = AssertSubscriber.create();
+
+			sink.next(1).complete();
+			flux.subscribe(first);
+
+			first.assertNoValues().assertComplete();
+		}
+
+		@Test
+		public void immediatelyErrorFirstSubscriber() {
+			AssertSubscriber<Integer> first = AssertSubscriber.create();
+
+			sink.next(1).error(new IllegalStateException("boom"));
+			flux.subscribe(first);
+
+			first.assertNoValues().assertErrorMessage("boom");
+		}
+
+		@Test
+		public void immediatelyCompleteLateSubscriber() {
+			flux.subscribe(); //first subscriber
+			AssertSubscriber<Integer> late = AssertSubscriber.create();
+
+			sink.next(1).complete();
+			flux.subscribe(late);
+
+			late.assertNoValues().assertComplete();
+		}
+
+		@Test
+		public void immediatelyErrorLateSubscriber() {
+			flux.onErrorReturn(-1).subscribe(); //first subscriber, ignore errors
+			AssertSubscriber<Integer> late = AssertSubscriber.create();
+
+			sink.next(1).error(new IllegalStateException("boom"));
+			flux.subscribe(late);
+
+			late.assertNoValues().assertErrorMessage("boom");
+		}
+	}
+
+	@Nested
+	class Multicast {
+
+		private Sinks.StandaloneFluxSink<Integer> sink;
+		private Flux<Integer> flux;
+
+		@BeforeEach
+		void createInstance() {
+			sink = Sinks.multicast();
+			flux = sink.asFlux();
+		}
+
+		@Test
+		void fluxViewReturnsSameInstance() {
```

Comment:
could use an abstract parent class that checks common behaviours like this one

Reply:
done: made a dynamic test generation to express more directly the semantics found in the javadocs